### PR TITLE
Share convert 2

### PR DESCRIPTION
**Pull request by bendecoste**

### CODE REVIEW

**Comment by bendecoste** [comment]
Requires CHANGELOG update

File: tf_encrypted/tensor/prime.py [hunk 1]
```diff
@@ -125,6 +129,9 @@ def _lift(x, y) -> Tuple[PrimeTensor, PrimeTensor]:
     if isinstance(x, int) and isinstance(y, PrimeTensor):
         return y.factory.tensor(np.array([x])), y
 
+    # if isinstance(x, PrimeTensor) and isinstance(y, OddImplicitTensor):
```
Comment (mortendahl) [inline tf_encrypted/tensor/prime.py]: left overs?

File: tf_encrypted/tensor/prime.py [hunk 2]
```diff
@@ -7,6 +7,7 @@
 
 from .factory import AbstractFactory, AbstractTensor, AbstractConstant, AbstractVariable, AbstractPlaceholder
 from .shared import binarize, im2col
+# from .odd_implicit import OddImplicitTensor
```
Comment (mortendahl) [inline tf_encrypted/tensor/prime.py]: left overs?

File: tf_encrypted/tensor/int64.py [hunk 3]
```diff
@@ -223,6 +224,29 @@ def squeeze(self, axis: Optional[List[int]]=None) -> 'Int64Tensor':
     def negative(self) -> 'Int64Tensor':
         return Int64Tensor(tf.negative(self.value))
 
+    def to_odd_modulus(self, factory: OddImplicitFactory) -> OddImplicitTensor:
```
Comment (mortendahl) [inline tf_encrypted/tensor/int64.py]: let's use casting instead, ie along the lines of https://github.com/mortendahl/tf-encrypted/blob/master/tf_encrypted/tensor/prime.py#L112-L113

File: tf_encrypted/tensor/int32.py [hunk 4]
```diff
@@ -177,6 +178,12 @@ def equal(self, other, factory: AbstractFactory=int32factory) -> 'AbstractTensor
         x, y = _lift(self, other)
         return factory.tensor(tf.cast(tf.equal(x.value, y.value), dtype=factory.native_type))
 
+    def compute_wrap(self, y: AbstractTensor, modulus: int) -> AbstractTensor:
+        return Int32Tensor(tf.cast(self.value + y.value >= modulus, dtype=tf.int32))
+
+    def to_odd_modulus(self, factory: OddImplicitFactory) -> OddImplicitTensor:
```
Comment (mortendahl) [inline tf_encrypted/tensor/int32.py]: see comment on casting elsewhere

File: tf_encrypted/tensor/int32.py [hunk 5]
```diff
@@ -177,6 +178,12 @@ def equal(self, other, factory: AbstractFactory=int32factory) -> 'AbstractTensor
         x, y = _lift(self, other)
         return factory.tensor(tf.cast(tf.equal(x.value, y.value), dtype=factory.native_type))
 
+    def compute_wrap(self, y: AbstractTensor, modulus: int) -> AbstractTensor:
+        return Int32Tensor(tf.cast(self.value + y.value >= modulus, dtype=tf.int32))
```
Comment (mortendahl) [inline tf_encrypted/tensor/int32.py]: won't this overflow if we're working with int32s?

File: tf_encrypted/protocol/securenn.py [hunk 6]
```diff
@@ -31,6 +32,7 @@ def __init__(
         server_2: Optional[Player] = None,
         prime_factory: Optional[AbstractFactory] = None,
         odd_factory: Optional[AbstractFactory] = None,
+        implicit_factory: Optional[AbstractFactory] = None,
```
Comment (mortendahl) [inline tf_encrypted/protocol/securenn.py]: I think `odd_factory` was intended for this!

File: tf_encrypted/protocol/securenn.py [hunk 7]
```diff
@@ -299,17 +306,104 @@ def equal_zero(self, x, out_dtype: Optional[AbstractFactory]=None):
...
+            np_0 = np_0.to_odd_modulus(self.implicit_factory)
+
+            np_odd = self._share_and_wrap(np_0, is_scaled=False)
+
+        # line 9
+        # n = np_odd + npp - np_odd * npp * 2
+
+        # line 10, j = 0
+        with tf.device(self.server_0.device_name):
+            n_0 = np_odd.share0 + npp.value_on_0 - (np_odd.share0 * npp.value_on_0 * 2)
```
Comment (mortendahl) [inline tf_encrypted/protocol/securenn.py]: this looks like an xor

File: tf_encrypted/protocol/securenn.py [hunk 8]
```diff
@@ -299,17 +306,104 @@ def equal_zero(self, x, out_dtype: Optional[AbstractFactory]=None):
...
+            a_0 = a.share0.to_odd_modulus(self.implicit_factory)
+            theta_0 = beta_0.to_odd_modulus(self.implicit_factory) + (((-alpha).to_odd_modulus(self.implicit_factory) - 1) * 1) + gamma.share0 + n_0
```
Comment (mortendahl) [inline tf_encrypted/protocol/securenn.py]: I think you should be able to compute `theta` without using `tf.device`

File: tf_encrypted/protocol/securenn.py [hunk 9]
```diff
@@ -299,17 +306,104 @@ def equal_zero(self, x, out_dtype: Optional[AbstractFactory]=None):
...
+
+            y_1 = a_1 - theta_1
+
+        # line 11
+        y = PondPrivateTensor(self, y_0, y_1, is_scaled=False)
+        return y
+
+    def compute_wrap(self, s0, s1, L: int):
+        # classical overflow
+        overflow_max = tf.cast(tf.logical_and((s0.value > 0), (s1.value > tf.int64.max - s0.value)), dtype=tf.bool)
```
Comment (mortendahl) [inline tf_encrypted/protocol/securenn.py]: you should be able to get `int64` from the factory so that this is not hardcoded

File: tests/test_share_convert.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            'crypto_producer'
+        ])
+
+    def test_share_convert(self):
+
+        prot = tfe.protocol.SecureNN(
+            tensor_factory=int64factory
+        )
+
+        val_a = np.array([100, -100])
```
Comment (mortendahl) [inline tests/test_share_convert.py]: could we try with a bunch of random values instead? i'm worried this won't catch corner cases

File: tf_encrypted/protocol/securenn.py (hunk 8 above)
Comment (bendecoste) [reply tf_encrypted/protocol/securenn.py]: I think we'd need to write casting for pond tensors which I can do but might be a little out of scope here.

**Comment by bendecoste** [comment]
@mortendahl I addressed most PR comments or replied to them.

I actually uncovered what I think is a bug.  If you try to share convert a sufficiently large number it will fail ~5% of the time.  I changed the test to show this.  Not sure it's worth fixing right now because we are thinking that we can bypass share convert anyway.  Considering just deferring to a different issue?

**Comment by mortendahl** [comment]
believe we can close this guy per https://github.com/mortendahl/tf-encrypted/pull/306

**Comment by mortendahl** [comment]
@bendecoste where did we end up re failing 5% of the time?

**Comment by mortendahl** [comment]
any thoughts @bendecoste?

**Comment by mortendahl** [comment]
closing since we don't need share convert for current applications and errors was never solved
